### PR TITLE
[MRG+1] Update base estimator

### PR DESCRIPTION
**Pull request by Eric89GXL**
Taking over #3988, which was a takeover of #3570.

@kingjr can you see if this satisfies your case?

### CODE REVIEW

File: mne/decoding/transformer.py [hunk 1]
```diff
@@ -79,10 +79,16 @@ def fit(self, epochs_data, y):
...
         for key, this_pick in picks_list.items():
             if self.with_mean:
-                ch_mean = X[:, this_pick, :].mean(axis=1)[:, None, :]
+                if len(this_pick) == 0:
+                    ch_mean = np.nan * np.ones((X.shape[0], 1, X.shape[2]))
+                else:
+                    ch_mean = X[:, this_pick, :].mean(axis=1, keepdims=True)
                 self.ch_mean_[key] = ch_mean  # TODO rename attribute
             if self.with_std:
-                ch_std = X[:, this_pick, :].mean(axis=1)[:, None, :]
```
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: @kingjr I just noticed this error in `Scaler.fit`. This is bug (setting `self.ch_std_[key]` to the *mean*), right?

Comment (kingjr) [inline mne/decoding/transformer.py]: Mmmh, that seems weird, but I'm not sure. I think scaler is supposed to take the std across time samples and epochs for each channel separately.

@jasmainak ?

**Comment by kingjr** [comment]
yes

**Comment by Eric89GXL** [comment]
Okay, I'll update `whats_new.rst` about a bug. This is ready for review whenever you have time

**Comment by kingjr** [comment]
Can you add a test?

I'm sorry, the next two weeks will be very busy for me, I don't think I can review PRs seriously.

**Comment by Eric89GXL** [comment]
`BaseEstimator` is implicitly tested, as it's a base class of a bunch of `Decoding` things -- it's also copy-paste from `sklearn` to `fixes.py` so I don't think we need to test that. Do you mean test the `std` thing?

File: mne/decoding/transformer.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: Right now both are taken across channels, not time...

**Comment by kingjr** [comment]
yes the std error

**Comment by Eric89GXL** [comment]
Well the docstring does say "standardize data across channels", but that's a bit vague. I would expect this to mean each channel (over time) would then have mean 0 std 1, but what it does is instead makes each time point (over channels) have mean 0 std 1. @kingjr which did you expect?

**Comment by Eric89GXL** [comment]
@jasmainak do you have time to look? I clarified docs and added a test.

**Comment by kingjr** [comment]
I would expect :

```python
X = epochs.get_data()
X.transpose(1, 0, 2).reshape(X.shape[1], -1)  # stack epochs and time
together
X -= X.mean(1) # mean over time & epochs
X /= X.std(1) # std over time and epochs
X.reshape(n_chans, n_times, n_trials).transpose(2, 0, 1)  # back to
expected dimension
```

File: mne/decoding/transformer.py (hunk 1 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: it's indeed a bug, but I thought it was already fixed by @kaichogami 

It was meant to standardize separately for each channel type ...

File: mne/decoding/transformer.py [hunk 2]
```diff
@@ -79,10 +83,16 @@ def fit(self, epochs_data, y):
 
         for key, this_pick in picks_list.items():
             if self.with_mean:
-                ch_mean = X[:, this_pick, :].mean(axis=1)[:, None, :]
+                if len(this_pick) == 0:
+                    ch_mean = np.nan * np.ones((X.shape[0], 1, X.shape[2]))
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: why this? You can no beforehand if a channel type is present or not

File: mne/decoding/tests/test_transformer.py [hunk 3]
```diff
@@ -42,17 +42,16 @@ def test_scaler():
...
-        X = scaler.fit_transform(epochs_data, y)
-        assert_true(X.shape == epochs_data.shape)
-        X2 = scaler.fit(epochs_data, y).transform(epochs_data)
-
+    X = scaler.fit_transform(epochs_data, y)
+    assert_true(X.shape == epochs_data.shape)
+    X2 = scaler.fit(epochs_data, y).transform(epochs_data)
     assert_array_equal(X2, X)
+    # these should be across time
+    assert_allclose(X.std(-2), 1.)
```
Comment (jasmainak) [inline mne/decoding/tests/test_transformer.py]: Could you add `axis=-2` if you don't mind ...

**Comment by Eric89GXL** [comment]
@kingjr that's not what it does -- not even on a per-epoch basis. It normalizes each time point, within each epoch, to have mean zero std 1 (for each channel type). Have you been using it expecting the other way?

File: mne/decoding/transformer.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: `self.std_` is a public attribute that always has a value for `mag`, `grad`, and `eeg`. This gives it the same value it was getting before, but does it without taking the mean or std of an empty slice

File: mne/decoding/transformer.py [hunk 4]
```diff
@@ -17,7 +18,10 @@
 
 
 class Scaler(TransformerMixin):
-    """Standardize data across channels.
+    u"""Standardize data across channels.
+
+    By default, this makes each time point (within each epoch) have
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: hmm ... but that's not what it does. `StandardScaler` in `scikit-learn` already lets you do that

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: That's what the *math* does. What is it supposed to do?

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: ... at least on a channel-type-by-channel-type basis that's what it does

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: `epochs_data : array, shape (n_epochs, n_channels, n_times)`

and the `mean` and `std` are taken along `axis=1`, i.e. the channels

Comment (jasmainak) [inline mne/decoding/transformer.py]: Yes, each channel type has a different scale ... the purpose of `Scaler` was to bring them all to the same scale before decoding

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: Right, this makes it so that each time point, within each epoch, have mu=0 std=1 *for each channel type*. Is that the intent? Then all I need to do is update the docstring to say "within each channel type" or so.

It's a bit weird, though -- it seems like you'd want to operate across all time points, or across all epochs, or so... I would have naively expected at least the scaling portion to just compute the std across all times and epochs for each channel type, and scale all the data for each channel type by that value.

File: mne/decoding/transformer.py (hunk 2 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: I see ... but you can `'mag' in info` to know if the dict should have all the keys? Not sure if it will break existing scripts ...

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: If you're confident nobody is really using the dict then yes it can be simplified

File: mne/decoding/transformer.py (hunk 4 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: > It's a bit weird, though -- it seems like you'd want to operate across all time points, or across all epochs, or so... I would have naively expected at least the scaling portion to just compute the std across all times and epochs for each channel type, and scale all the data for each channel type by that value.

Indeed ... you are correct. Do you want to change the behavior as it is now? You have my +1 if you want to make it like that

Comment (kingjr) [inline mne/decoding/transformer.py]: I think this is a bug. The aim of the scaler is originally to make the EEG/grad and mag varying around the same scale. It should remove the channel mean and divide by its deviation over time samples (i.e. all time points => meaning time * epochs). 

This can change significantly some results e.g. `make_pipeline(Scaler(), Xdawn())`

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: So the updated docstring would be something like:
```
    By default, this makes each time point (within each epoch) have
    μ=0, σ=1 for each channel type (``'mag', 'grad', 'eeg'``) independently.
```
And I would have expected something more like:
```
   By default, this makes each channel have zero mean across time,
   and scales all channels of the same type (``'grad', 'mag', 'eeg'``)
   by the same factor to achieve unit variance across all time points
   within all epochs.
```
I'm also a bit surprised this is really needed, given that e.g. in our `cov` code we do simple `dict`-based scaling and it seems to work well enough. Why not use that approach here?

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: @kingjr @jasmainak how about a `dict`-based method instead of using empirical `std`?

Comment (jasmainak) [inline mne/decoding/transformer.py]: > It should remove the channel mean and divide by its deviation over time samples (i.e. all time points => meaning time * epochs)

The problem I see with this approach is if there are artifactual epochs, your std dev won't be accurate. You have to use a robust measure ...

Comment (kingjr) [inline mne/decoding/transformer.py]: @jasmainak This is an independent issue IMO. 

#3672 is supposed to take care of this by using robust scaler instead of the z-score.

I suggest we move this discussion & correction to #3672; this PR is about updating the BaseEstimator.

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: > The problem I see with this approach is if there are artifactual epochs, your std dev won't be accurate. You have to use a robust measure ...

I think this is part of the motivation for using `scalings=dict(...)` in `cov` computation. But there being within a couple of orders of magnitude is enough. Here maybe it's more sensitive.

Anyway, back to `std` -- maybe trimmed standard deviation, then?

Comment (jasmainak) [inline mne/decoding/transformer.py]: > Anyway, back to std -- maybe trimmed standard deviation, then?

Yeah, I don't know for real what works and what doesn't ... there is also [median absolute deviation](https://en.wikipedia.org/wiki/Median_absolute_deviation)

Comment (jasmainak) [inline mne/decoding/transformer.py]: > #3672 is supposed to take care of this by using robust scaler instead of the z-score.

Indeed, this is what I thought. I didn't realize the PR wasn't merge and it came up in a diff in this PR!

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: > I suggest we move this discussion & correction to #3672; this PR is about updating the BaseEstimator.

Agreed. I'll keep the docstring and warning fixes in here, and let the "real" fixes be done in another PR.

File: mne/decoding/transformer.py (hunk 2 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: okay maybe later :)

**Comment by Eric89GXL** [comment]
In that case, this one should be good to go. Updated `axis=-2` in the tests to reflect current behavior.

**Comment by jasmainak** [comment]
LGTM ... maybe one more person can take a look what's up with the `BaseEstimator`? I didn't check that carefully ...

**Comment by Eric89GXL** [comment]
@agramfort this could use another set of eyes and merge if you have time

**Comment by agramfort** [comment]
@kingjr @dengemann you merged this Scaler code but it seems that it's not used in any example or tutorial and to be honest I am not sure what's the use case. You use StandardScaler from sklearn in many places to standardize across epochs which is classic for ML but this Scaler approach I don't know. 

@kingjr @dengemann or maybe @jasmainak can you help me understand?

**Comment by kingjr** [comment]
Yes, we need to add examples.

This is useful when doing decoding pipelines that use multiple time points
of with mixed channels. E.g. Xdawn(MEEG)

**Comment by Eric89GXL** [comment]
In the meantime this should be good to go, for Scaler it only clarifies
current behavior

**Comment by agramfort** [comment]
>
> This is useful when doing decoding pipelines that use multiple time points
> of with mixed channels. E.g. Xdawn(MEEG)
>

can you share a code snippet to illustrate the benefit / necessity?
I would just have just scaled each channel type by a scalar.

**Comment by kingjr** [comment]
You'd mainly need a data simulation snippet than an analytic one. E.g. (not
tested)

```python
# make up random MEEG data
X = np.random.randn(n_epochs, n_chans, n_times)
y = np.arange(n_epochs) % 2

# Scale by chan type
X[:, :306:3, :] /= 1e-11  # scale grad channels
X[:, 1:306:3, :] /= 1e-11  # scale grad channels
X[:, 2:306, :] /= 1e-13  # scale mag channels
X[:, 306:, :] /= 1e6 # scale eeg channels

# Add annoying typicalities
baseline_shifts = np.random.randn(n_epochs, n_chans)
X += baseline_shits[:, :, np.newaxis]
bad_channels = np.random.randint(0, n_channels, 10)
X[:, bad_channels, :] = np.random.randn(n_epochs, 10, n_times]

# Pipeline
train, test = range(n_epochs//2), range(n_epochs//2, n_epochs)
xdawn = make_pipeline(mne.decoding.Scaler(), XdawnTransformer())
xdawn.fit(X[train], y[train])
X_clean = xdawn.transform(X[test]).inverse_transform(X[test])

```

**Comment by agramfort** [comment]
>
> ```python
> # make up random MEEG data
> X = np.random.randn(n_epochs, n_chans, n_times)
> y = np.arange(n_epochs) % 2
>
> # Scale by chan type
> X[:, :306:3, :] /= 1e-11 # scale grad channels
> X[:, 1:306:3, :] /= 1e-11 # scale grad channels
> X[:, 2:306:3, :] /= 1e-13 # scale mag channels
> X[:, 306:, :] /= 1e6 # scale eeg channels
>

this suggests a global scaling per channel type as I said above.


> # Add annoying typicalities
> baseline_shifts = np.random.randn(n_epochs, n_chans)
> X += baseline_shits[:, :, np.newaxis]
>

this will get fixed by high pass or baseline correction


> bad_channels = np.random.randint(0, n_channels, 10)
> X[:, bad_channels, :] = np.random.randn(n_epochs, 10, n_times]
>

Scaler will not be robust to bad channels of if artifact is big
it might wash out your effect (div by std)

**Comment by kingjr** [comment]
I'm not the creator of this class, I'm just indicating where it can be
useful. My and Asish' contributions were made to make it API compatible
with the other decoding classes and with scikit-learn's pipeline.

> this suggests a global scaling per channel type as I said above.

yes, what else?

> this will get fixed by high pass or baseline correction

which can introduce other issues; that's another discussion.

> Scaler will not be robust to bad channels of if artifact is big; it might
wash out your effect (div by std)

Yes, we already have had this discussion in multiple places. This is why I
suggested to add a method='robust'.

**Comment by jasmainak** [comment]
@agramfort the fix to `Scaler` is bugfix and documentation. The intended behavior has not been changed. I can summarize the discussion IRL for you if you want.

Everyone seems to agree that a constant scaler is the way to go. I think I might have misunderstood what needed to be implemented 4 years ago and it passed through the reviews unnoticed back then ... it can be fixed but this is not the PR to fix it. It's https://github.com/mne-tools/mne-python/pull/3672

**Comment by agramfort** [comment]
ok fair enough. thanks @jasmainak for clarifying. I had not realized it the history...

let's discuss this IRL during the sprint
